src/Foundatio.AWS/Queues/SQSQueue.cs

```diff
@@ -0,0 +1,269 @@
+                QueueUrl = _queueUrl,
+                ReceiptHandle = entry.UnderlyingMessage.ReceiptHandle,
+            };
+
+            await _client.Value.DeleteMessageAsync(request).ConfigureAwait(false);
+
+            Interlocked.Increment(ref _completedCount);
+            queueEntry.MarkCompleted();
+            await OnCompletedAsync(queueEntry).ConfigureAwait(false);
+
+        }
+
+        public override async Task AbandonAsync(IQueueEntry<T> queueEntry) {
+            var entry = ToQueueEntry(queueEntry);
+
+            // re-queue and wait for processing
+            var visibilityTimeout = Convert.ToInt32(_workItemTimeout.TotalSeconds);
+            var request = new ChangeMessageVisibilityRequest {
+                QueueUrl = _queueUrl,
+                VisibilityTimeout = visibilityTimeout,
+                ReceiptHandle = entry.UnderlyingMessage.ReceiptHandle,
+            };
+
+            await _client.Value.ChangeMessageVisibilityAsync(request).ConfigureAwait(false);
+
+            Interlocked.Increment(ref _abandonedCount);
+            queueEntry.MarkAbandoned();
+            await OnAbandonedAsync(queueEntry).ConfigureAwait(false);
+        }
+
+        protected override Task<IEnumerable<T>> GetDeadletterItemsImplAsync(CancellationToken cancellationToken) {
+            throw new NotImplementedException();
+        }
+
+        protected override async Task<QueueStats> GetQueueStatsImplAsync() {
+            var attributeNames = new List<string> { QueueAttributeName.ApproximateNumberOfMessages };
+            var request = new GetQueueAttributesRequest(_queueUrl, attributeNames);
+
+            var queueAttributes = await _client.Value.GetQueueAttributesAsync(request).ConfigureAwait(false);
+
+            return new QueueStats {
+                Queued = queueAttributes.ApproximateNumberOfMessages,
+                Working = 0,
+                Deadletter = 0,
+                Enqueued = _enqueuedCount,
+                Dequeued = _dequeuedCount,
+                Completed = _completedCount,
+                Abandoned = _abandonedCount,
+                Errors = _workerErrorCount,
+                Timeouts = 0
+            };
+        }
+
+        public override async Task DeleteQueueAsync() {
+            if (String.IsNullOrEmpty(_queueUrl))
+                return;
+
+            var request = new DeleteQueueRequest(_queueUrl);
+            var response = await _client.Value.DeleteQueueAsync(request).ConfigureAwait(false);
+
+            _enqueuedCount = 0;
+            _dequeuedCount = 0;
+            _completedCount = 0;
+            _abandonedCount = 0;
+            _workerErrorCount = 0;
+        }
+
+        protected override void StartWorkingImpl(Func<IQueueEntry<T>, CancellationToken, Task> handler, bool autoComplete, CancellationToken cancellationToken) {
+            Task.Run(async () => {
+                while (!cancellationToken.IsCancellationRequested) {
```

Comment:
we use the linked cancellation token here too, otherwise this task might never end (especially in unit tests) when the test is disposed

https://github.com/exceptionless/Foundatio/blob/master/src/Foundatio.Redis/Queues/RedisQueue.cs#L217